druid-shell/src/platform/x11/clipboard.rs

```diff
@@ -32,26 +75,193 @@ impl Clipboard {
-        // TODO(x11/clipboard): implement Clipboard::preferred_format
-        warn!("Clipboard::preferred_format is currently unimplemented for X11 platforms.");
-        None
+    pub fn preferred_format(&self, formats: &[FormatId]) -> Option<FormatId> {
+        let available = self.available_type_names();
+        formats
+            .iter()
+            .find(|f1| available.iter().any(|f2| *f1 == f2))
+            .copied()
     }
 
-    pub fn get_format(&self, _format: FormatId) -> Option<Vec<u8>> {
-        // TODO(x11/clipboard): implement Clipboard::get_format
-        warn!("Clipboard::get_format is currently unimplemented for X11 platforms.");
-        None
+    pub fn get_format(&self, format: FormatId) -> Option<Vec<u8>> {
+        self.do_transfer(format, |prop| prop.value)
     }
 
+    #[allow(clippy::needless_collect)]
     pub fn available_type_names(&self) -> Vec<String> {
-        // TODO(x11/clipboard): implement Clipboard::available_type_names
-        warn!("Clipboard::available_type_names is currently unimplemented for X11 platforms.");
-        vec![]
+        let requests = self
+            .do_transfer("TARGETS", |prop| {
+                prop.value32()
+                    .map(|iter| iter.collect())
+                    .unwrap_or_default()
+            })
+            .unwrap_or_default()
+            .into_iter()
+            .filter_map(|atom| self.connection.get_atom_name(atom).ok())
+            .collect::<Vec<_>>();
+        // We first send all requests above and then fetch the replies with only one round-trip to
+        // the X11 server. Hence, the collect() above is not unnecessary!
+        requests
+            .into_iter()
+            .filter_map(|req| req.reply().ok())
+            .filter_map(|reply| String::from_utf8(reply.name).ok())
+            .collect()
+    }
+
+    fn do_transfer<R, F>(&self, format: FormatId, converter: F) -> Option<Vec<R>>
+    where
+        R: Clone,
+        F: FnMut(GetPropertyReply) -> Vec<R>,
+    {
+        match self.do_transfer_impl(format, converter) {
+            Ok(result) => result,
+            Err(error) => {
+                warn!("Error in Clipboard::do_transfer: {:?}", error);
+                None
+            }
+        }
+    }
+
+    fn do_transfer_impl<R, F>(
+        &self,
+        format: FormatId,
+        mut converter: F,
+    ) -> Result<Option<Vec<R>>, ReplyOrIdError>
+    where
+        R: Clone,
+        F: FnMut(GetPropertyReply) -> Vec<R>,
+    {
+        debug!("Getting clipboard contents in format {}", format);
+
+        let conn = &*self.connection;
+        let (format_atom, clipboard_atom, incr_atom) = {
+            let format = conn.intern_atom(false, format.as_bytes())?;
+            let clipboard = conn.intern_atom(false, b"CLIPBOARD")?;
+            let incr = conn.intern_atom(false, b"INCR")?;
+            (
+                format.reply()?.atom,
+                clipboard.reply()?.atom,
+                incr.reply()?.atom,
```

Comment:
if we delaying this reply after window creation, is it worth the less readability?

Reply:
Uhm. What exactly do you mean with "is it worth"? What's the benefit?

I guess you want to understand a round-trip to the X11 server (and I applaud that goal - I sometimes feel like I am the only one caring about that). However, I don't see the reduced number of round trips. There is one round trip getting the result for the `InternAtom` requests and another round trip waiting for the result of `ConvertSelection`. You suggestions just moves a `CreateWindow` request from the second to the first round-trip.

Why does that make a difference? Or am I misunderstanding you?

(I think I applied fixes for the other comments. Thanks for those comments!)

Reply:
Sorry I don't quite understand x11rb, I thought as a rule of thumb: the later `.reply()`, the better. Thanks for explaining :)

Reply:
> I thought as a rule of thumb: the later .reply(), the better.

Well, yeah. That's the short version. Let me try to write down a long version:

When you send a request (e.g. `conn.intern_atom(...)`), the request is serialised and its bytes are put into an output buffer. The next request is also put into the output buffer etc. This buffer is actually sent to the X11 server in three cases:
- `conn.flush()` is called
- A request is sent but the buffer is full (for example because of lots of previous `intern_atom()` calls, but this can also happen with a big image in a `put_image()` call). In this case, all or parts of the new request can again end up in the buffer.
- You call `.reply()` for a request that is still in the output buffer.

Next part is `.reply()`: This function, well, waits for the reply to a request! There is again a buffer (actually a `VecDeque`, I think) of replies that were already read from the socket but not yet returned from API functions. When you call `.reply()`, it first checks this buffer. If the right reply is not found, the code waits for a new packet from the X11 server and then checks the queue again.

-------

With the above in mind, we can look at code like this:
```
let foo = conn.intern_atom("foo")?.reply()?;
let bar = conn.intern_atom("bar")?.reply()?;
```
The first line puts a request into the output buffer (`intern_atom()`), and then immediately flushes the output buffer and waits for a reply from the X11 server. This "I just sent a request and now wait for its reply" is what I call a round trip since when you measure the time it takes to execute the code, these round-trips are what take most of the time.

The second line does the same again and again takes some time.

With code like this instead:
```
let foo_cookie = conn.intern_atom("foo")?;
let bar_cookie = conn.intern_atom("bar")?;
let foo = foo_cookie.reply()?;
let bar = bar_cookie.reply()?;
```
This code will send both requests in `foo_cookie.reply()`, because both are already in the output buffer. It now has to wait for the reply, but the X11 server basically gets both requests "together".

In `bar_cookie.reply()`, the X11 connection will not yet actually have `read()` the reply from the X11 connection, since the reply for "foo" comes in first and the code stops reading from the connection once it got the right reply. Still, this will take less time.

----

As an extreme example, there is this code: https://github.com/psychon/x11rb/blob/5b9aafa515d2bc3b2c7fd8f1bfbd01cc5dff4431/examples/tutorial.rs#L203-L239

This code will first do lots of `conn.intern_atom().reply()` and print how long this took. Then it does the same thing again, but with first doing all the `intern_atom()` calls so that they all end up in the output buffer, and afterwards doing all the `reply()` calls. Thus, the second version only needs one round trip while the first one does 500 round trips.

The output of this code with `--release`:
```
bad use time:  5.812688ms
good use time: 970.326µs
ratio:         5.990448570892669
```
The output without `--release`:
```
bad use time:  16.929757ms
good use time: 3.958736ms
ratio:         4.276556203798384
```
So, yeah... avoiding round trips is a bit faster, but this example needs 500 round trips to make a measurable difference. However, I bet that if I used ssh X11 forwarding (´ssh -X`) over the internet, the round-trip-time to the X11 server would go up and the difference would be higher. Dunno if anyone still uses this forwarding and if one should care a lot about it when coding...

----

Now for the code in this PR: Actually... no, I think my original comment says enough about this. It's still the same number of round-trips ("send a request and wait for its reply"). I hope that this comment explains everything a bit better and makes it clearer what is going on.